src/funnels/Depositor.sol

```diff
@@ -0,0 +1,275 @@
+        bytes calldata data
+    ) external {
+        MintCallbackData memory decoded = abi.decode(data, (MintCallbackData));
+        address pool = address(_getPool(decoded.gem0, decoded.gem1, decoded.fee));
+        require(msg.sender == pool, "Depositor/sender-not-a-pool");
+
+        if (amt0Owed > 0) GemLike(decoded.gem0).transferFrom(buffer, msg.sender, amt0Owed);
+        if (amt0Owed > 0) GemLike(decoded.gem1).transferFrom(buffer, msg.sender, amt1Owed);
+    }
+
+    struct LiquidityParams {
+        address gem0;
+        address gem1;
+        uint24  fee;
+        int24   tickLower;
+        int24   tickUpper;
+        uint128 liquidity;   // Useful for clearing out the entire liquidity of a position
+        uint256 amt0Desired; // relevant only if liquidity == 0
+        uint256 amt1Desired; // relevant only if liquidity == 0
+        uint256 amt0Min;
+        uint256 amt1Min;
+    }
+
+    function deposit(LiquidityParams memory p)
+        external
+        auth
+        returns (uint128 liquidity, uint256 amt0, uint256 amt1)
+    {
+        require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
+
+        PairLimit memory limit = limits[p.gem0][p.gem1];
+        require(block.timestamp >= limit.zzz + limit.hop, "Depositor/too-soon");
+        limits[p.gem0][p.gem1].zzz = uint64(block.timestamp);
+
+        UniV3PoolLike pool = _getPool(p.gem0, p.gem1, p.fee);
+        liquidity = (p.liquidity == 0)
+            ? _getLiquidityForAmts(pool, p.tickLower, p.tickUpper, p.amt0Desired, p.amt1Desired)
+            : p.liquidity;
+
+        (amt0, amt1) = pool.mint({
+            recipient: address(this),
+            tickLower: p.tickLower,
+            tickUpper: p.tickUpper,
+            amount   : liquidity,
+            data     : abi.encode(MintCallbackData({gem0: p.gem0, gem1: p.gem1, fee: p.fee}))
+        });
+        require(amt0 >= p.amt0Min && amt1 >= p.amt1Min, "Depositor/exceeds-slippage");
+        require(amt0 <= limit.cap0 && amt1 <= limit.cap1, "Depositor/exceeds-cap");
+
+        emit Deposit(msg.sender, p.gem0, p.gem1, liquidity, amt0, amt1);
+    }
+
+    function withdraw(LiquidityParams memory p, bool takeFees)
+        external
+        auth
+        returns (uint128 liquidity, uint256 amt0, uint256 amt1)
+    {
+        require(p.gem0 < p.gem1, "Depositor/wrong-gem-order");
+
+        PairLimit memory limit = limits[p.gem0][p.gem1];
+        require(block.timestamp >= limit.zzz + limit.hop, "Depositor/too-soon");
+        limits[p.gem0][p.gem1].zzz = uint64(block.timestamp);
+
+        UniV3PoolLike pool = _getPool(p.gem0, p.gem1, p.fee);
+        liquidity = (p.liquidity == 0)
+            ? _getLiquidityForAmts(pool, p.tickLower, p.tickUpper, p.amt0Desired, p.amt1Desired)
+            : p.liquidity;
+
+        (amt0, amt1) = pool.burn({ tickLower: p.tickLower, tickUpper: p.tickUpper, amount: liquidity });
+        require(amt0 >= p.amt0Min  && amt1 >= p.amt1Min,  "Depositor/exceeds-slippage");
```

Comment:
```suggestion
        require(amt0 >= p.amt0Min && amt1 >= p.amt1Min,  "Depositor/exceeds-slippage");
```
Otherwise add the space in `deposit` to mirror this.